cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -1520,27 +1530,264 @@ ReconciliationState createReconciliationState(Reconciliation reconciliation, Kaf
+         * are not used anymore by the pods.
+         *
+         * @return
+         */
+        Future<ReconciliationState> zkPersistentClaimDeletion() {
+            Future<ReconciliationState> futureResult = Future.future();
+            Future<List<PersistentVolumeClaim>> futurePvcs = pvcOperations.listAsync(namespace, Labels.fromMap(zkCluster.getSelectorLabels()));
+
+            futurePvcs.setHandler(res -> {
+                if (res.succeeded() && res.result() != null)    {
+                    List<String> maybeDeletePvcs = res.result().stream().map(pvc -> pvc.getMetadata().getName()).collect(Collectors.toList());
+                    List<String> desiredPvcs = zkCluster.generatePersistentVolumeClaims().stream().map(pvc -> pvc.getMetadata().getName()).collect(Collectors.toList());
+
+                    persistentClaimDeletion(maybeDeletePvcs, desiredPvcs).setHandler(futureResult.completer());
+                } else {
+                    futureResult.fail(res.cause());
+                }
+            });
+
+            return futureResult;
+        }
+
+        /**
+         * Deletion of PVCs after the cluster is deleted is handled by owner reference and garbage collection. However,
+         * this would not help after scale-downs. Therefore we check if there are any PVCs which should not be present
+         * and delete them when they are.
+         *
+         * This should be called only after the Statefulset reconciliation, rolling update and scale-down when the PVCs
+         * are not used anymore by the pods.
+         *
+         * @return
+         */
         Future<ReconciliationState> kafkaPersistentClaimDeletion() {
-            return persistentClaimDeletion(kafkaCluster.getStorage(), kafkaCluster.getReplicas(),
-                (storage, i) -> {
-                    String name = ModelUtils.getVolumePrefix(storage.getId());
-                    return name + "-" + KafkaCluster.kafkaClusterName(reconciliation.name()) + "-" + i;
-                });
+            Future<ReconciliationState> futureResult = Future.future();
+            Future<List<PersistentVolumeClaim>> futurePvcs = pvcOperations.listAsync(namespace, Labels.fromMap(kafkaCluster.getSelectorLabels()));
+
+            futurePvcs.setHandler(res -> {
+                if (res.succeeded() && res.result() != null)    {
+                    List<String> maybeDeletePvcs = res.result().stream().map(pvc -> pvc.getMetadata().getName()).collect(Collectors.toList());
+                    List<String> desiredPvcs = kafkaCluster.generatePersistentVolumeClaims(kafkaCluster.getStorage()).stream().map(pvc -> pvc.getMetadata().getName()).collect(Collectors.toList());
+
+                    persistentClaimDeletion(maybeDeletePvcs, desiredPvcs).setHandler(futureResult.completer());
+                } else {
+                    futureResult.fail(res.cause());
+                }
+            });
+
+            return futureResult;
+        }
```

Comment:
the above two methods are quite the same, do you think there is no way for having just one acting on K or ZK provided somehow as parameter?

Reply:
I do not think it is so easy, mainly because the `generatePersistentVolumeClaims` signatures are different for Kafka and for ZK. This is because it already expects that the Kafka storage with JBOD might allow to add or remove volumes (which does not apply for ZK since there is no JBD storage supported). So I would probably end up with something like `instanceof` which would be IMHO quite ugly